targetcli/ui_backstore.py

```diff
@@ -526,17 +529,45 @@ def __init__(self, parent):
         self.so_cls = UIBlockStorageObject
         UIBackstore.__init__(self, 'block', parent)
 
+    def _ui_block_ro_check(self, dev):
+        BLKROGET=0x0000125E
+        try:
+            f = os.open(dev, os.O_RDONLY)
+        except (OSError, IOError):
+            raise ExecutionError("Could not open %s" % dev)
+        # ioctl returns an int. Provision a buffer for it
+        buf = array.array('c', [chr(0)] * 4)
+        try:
+            fcntl.ioctl(f, BLKROGET, buf)
+        except (OSError, IOError):
+            os.close(f)
+            return False
+
+        os.close(f)
+        if struct.unpack('I', buf)[0] == 0:
+            return False
+        return True
+
     def ui_command_create(self, name, dev, readonly=None, wwn=None):
         '''
         Creates an Block Storage object. I{dev} is the path to the TYPE_DISK
         block device to use.
         '''
         self.assert_root()
 
-        readonly = self.ui_eval_param(readonly, 'bool', False)
+        ro_string = self.ui_eval_param(readonly, 'string', None)
+        if ro_string == None:
+            # attempt to detect block device readonly state via ioctl
+            readonly = self._ui_block_ro_check(dev)
+        else:
+            readonly = self.ui_eval_param(readonly, 'bool', False)
+
         wwn = self.ui_eval_param(wwn, 'string', None)
 
-        so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
+        try:
+            so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
+        except IOError:
```

Comment:
It was not clear why this change is needed. Which call in the BlockStorageObject constructor raises it? Is it set_udev_path?

Why also only catch this one here and not also catch the possible RTSLibError here too or have the higher level catch it with all exceptions?

Reply:
It's the rtslib backstore->enable I/O that triggers IOError: [Errno 13]
```
  File "/usr/lib/python2.7/site-packages/targetcli/ui_backstore.py", line 557, in ui_command_create
    so = BlockStorageObject(name, dev, readonly=readonly, wwn=wwn)
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 703, in __init__
    self._configure(dev, wwn, readonly)
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 720, in _configure
    self._enable()
  File "/usr/lib/python2.7/site-packages/rtslib_fb/tcm.py", line 158, in _enable
    fwrite(path, "1\n")
  File "/usr/lib/python2.7/site-packages/rtslib_fb/utils.py", line 79, in fwrite
    file_fd.write(str(string))
IOError: [Errno 13] Permission denied
```

Even with the patch to auto-detect block device read-only state, this can still be tripped if a read-only block device is provided alongside an explicit readonly=false, hence the need for this exception handler.

I'm not aware of other exceptions that can occur in this code path, so chose to only cater to the above IOError. I could change this to something more generic if you'd prefer.

Thanks everyone for the feedback so far.

Reply:
> I'm not aware of other exceptions that can occur in this code path, so chose to only cater to the above IOError. I could change this to something more generic if you'd prefer.

Every kernel backend module can return failure from their device configure callback so a write to the enable file can return failure from any of them, so I think we want to handle this in one place and not just fix one backend.

For example maybe the StorageObject class's _enable function can catch these errors and translate them to RTSLibErrors which are being caught already.
 
I think having one or two backend create functions catching it and the rest crashing is the worst option :) My patch
6b14c2af6790ad8073fcd4982363197f5c030de2
was the wrong way to go :) It should be reverted and use the generic code needed in this thread. I goofed up and did not know the other constrictors did not catch these errors :)

Reply:
Okay, I'll look at adding generic RTSLibErrors functionality. That said, I think the first commit could go in without the follow-up error handling change.

Reply:
Agree.